feat: improve max failed login mechanism

bradymiller opened the pull request:
fixes #5627 

Basically provides ability to set a timeout (from last failed login) where then ok to login.
Also sets default setting of 20 for failed logins (and 1 hour for the timeout above)
Displayed the failed user login counter and last failed date in the Users display.

Would it also be helpful to have ability to reset the failed login counter in the Users display?

bradymiller commented:
Prob also should make the last login fail date null as done in this prior fix:
https://github.com/openemr/openemr/pull/6198

stephenwaite commented:
![excited_elephant](https://user-images.githubusercontent.com/5618372/236203016-8c6e43e9-ee2a-4aa0-9b6f-ae8d840af3ba.gif)

bradymiller commented:
Added ability to reset counter from the user admin gui.

Also added mechanism to deal with failed logins from a ip address perspective (with ability to also flag ip address as always blocked and even flag to skip the prevent timing attack when always blocking an ip address (skipping the timing attack prevention will increase performance of server since don't need to run a fake hash for every attempt, but this means the bad actor will know that their ip address is being blocked which is good for one server, but in the big scheme of things, ie. the OpenEMR ecosystem of thousands of servers is actually better to not do this)

Working on:
Report to view the ip addresses and reset their counters and set the block and skip timing attack flags
Send email to admin when a user or ip address gets automatically blocked because of too many failed logins

bradymiller commented on interface/usergroup/usergroup_admin.php:
```diff
@@ -687,6 +703,26 @@ function authorized_clicked() {
                                 echo '</td>';
                             }
+                            if (AuthUtils::useActiveDirectory($iter["username"]) || empty($iter["active"])) {
+                                // LDAP bypasses failed login counter mechanism
```
remove above comment since this is no longer true

bradymiller commented on interface/usergroup/usergroup_admin.php:
```diff
@@ -687,6 +703,26 @@ function authorized_clicked() {
                                 }
                                 echo '</td>';
                             }
+                            if (AuthUtils::useActiveDirectory($iter["username"]) || empty($iter["active"])) {
```
Also remove the above check for useActiveDirectory since this is no longer true

bradymiller commented on interface/usergroup/usergroup_admin.php:
```diff
@@ -687,6 +703,26 @@ function authorized_clicked() {
+                                        echo ' (' . xlt('last on') . ' ' . text(oeFormatDateTime($queryCounter['last_login_fail'])) . ')';
+                                    }
+                                    if (AclMain::aclCheckCore('admin', 'super')) {
+                                        echo ' ' . '<button type="button" class="btn btn-sm btn-danger ml-1" onclick="resetCounter(' . attr_js($iter["username"]) . ')">' . xlt("Reset") . '</button>';
```
Change to "Reset Counter"

adunsulag commented on library/ajax/login_counter_ip_tracker.php:
```diff
@@ -0,0 +1,38 @@
+    CsrfUtils::csrfNotVerified(false);
+}
+
+if (!AclMain::aclCheckCore('admin', 'super')) {
```
Do you think frontend staff are going to need the ability to reset someone from being logged out?  This seems like a fairly common support staff function.  I understand the desire to lock it down to just an admin thing... but it seems like the support staff are going to hit this the most from someone.

bradymiller replied:
changed this to admin users acl (same that is needed to get to the admin users gui)

adunsulag commented on library/globals.inc.php:
```diff
@@ -2148,10 +2148,31 @@ function gblTimeZones()
+        'ip_max_failed_logins' => array(
+            xl('Maximum Failed Login Attempts From IP Address'),
+            'num',                            // data type
+            '100',                            // default
```
is the reason why you made this 100 due to a worry about NATed IP addresses?

bradymiller replied:
just picked that number. happy to go higher or lower on this (goal is to not break user installs since this will be default and also prevent bruteforce foolishness)

adunsulag commented on sql/7_0_1-to-7_0_2_upgrade.sql:
```diff
@@ -136,3 +136,24 @@ DROP TABLE IF EXISTS `pma_table_coords`;
+#IfNotTable ip_tracking
+CREATE TABLE `ip_tracking` (
+`id` bigint NOT NULL auto_increment,
+`ip_string` varchar(255) DEFAULT '',
```
This table will get large, does it make sense to store the ip string or store it as the 16 byte for ipv6 and 4byte for ipv4?  It will definitely be better for debugging/administration to have the string representation.

bradymiller replied:
note ip_string is not always a straightforward ip address. sometimes is 2 addresses in string form:
https://github.com/openemr/openemr/blob/9dc161d8d5b3fa71508b3e9f23b7c3e4aea49b97/library/sanitize.inc.php#L31

adunsulag commented on src/Common/Auth/AuthUtils.php:
```diff
@@ -943,35 +1005,90 @@ private function checkPasswordNotExpired($user)
+            if (!empty($GLOBALS['ip_time_reset_password_max_failed_logins']) &&
+                $GLOBALS['ip_time_reset_password_max_failed_logins'] > 0 &&
+                !empty($query['seconds_last_ip_login_fail']) &&
+                $query['seconds_last_ip_login_fail'] > $GLOBALS['ip_time_reset_password_max_failed_logins']
```
seems like you could end up with some wierd bugs if someone were to put a letter in the globals, would it be better to grab an intval of $GLOBALS['ip_time_reset_password_max_failed_logins'] and do the comparison that way?

bradymiller replied:
yep, prob good idea to go through the globals that using in this pr and ensure are int
TODO for me

adunsulag reviewed (commented):
Looked it over, and noted a couple things, overall though looks fine to me.

bradymiller commented:
User gui
![image](https://user-images.githubusercontent.com/278968/236764031-e7a57ce7-de4b-4316-8390-3fadd07f9df8.png)

ip tracking gui
![image](https://user-images.githubusercontent.com/278968/236764979-03cda9c7-4e44-43bb-8447-544b75889e0c.png)
